File: bin/favoringIsolatedMutability/scala/transactors/Account.scala
package com.agiledeveloper.pcj

import akka.transactor.Transactor
import akka.stm.Ref

//START:CODE
class Account extends Transactor {
  val balance = Ref(0)
  
  def atomically = {
    case Deposit(amount) =>
      if (amount > 0) {
        balance.swap(balance.get() + amount)
        println("Received Deposit request " + amount)
      }

    case Withdraw(amount) =>
      println("Received Withdraw request " + amount)
      if (amount > 0 && balance.get() >= amount) 
        balance.swap(balance.get() - amount)
      else {
        println("...insufficient funds...")
        throw new RuntimeException("Insufficient fund")
      }
 
    case FetchBalance =>
      self.replySafe(Balance(balance.get()))
  }
}
//END:CODE

File: src/concurrency-code/favoringIsolatedMutability/java/transactors/Account.java
/***
 * Excerpted from "Programming Concurrency on the JVM",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material, 
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose. 
 * Visit http://www.pragmaticprogrammer.com/titles/vspcon for more book information.
***/
package com.agiledeveloper.pcj;

import akka.transactor.UntypedTransactor;
import akka.stm.Ref;

public class Account extends UntypedTransactor {
  private final Ref<Integer> balance = new Ref<Integer>(0);
  
  public void atomically(final Object message) {
    if(message instanceof Deposit) {
      int amount = ((Deposit)(message)).amount;
      if (amount > 0) {
        balance.swap(balance.get() + amount);
        System.out.println("Received Deposit request " + amount);
      }
    }

    if(message instanceof Withdraw) {
      int amount = ((Withdraw)(message)).amount;
      System.out.println("Received Withdraw request " + amount);
      if (amount > 0 && balance.get() >= amount) 
        balance.swap(balance.get() - amount);
      else {
        System.out.println("...insufficient funds...");
        throw new RuntimeException("Insufficient fund");        
      }
    }
 
    if(message instanceof FetchBalance) {
      getContext().replySafe(new Balance(balance.get()));
    }
  }  
}
